### ISSUE TITLE
Vector Tile has POLYGON has ring with too few points.

### ISSUE THREAD

**Issue by Lee858** [systemed/tilemaker#143]
I’m converting data that contains polygons that represent public land ownership. The parcels are often large with inner polygons. The data has already been scrubbed for intersecting polygons, intersecting points, too few elements and small objects.  I’m using klokantech/tileserver-mapnik to view the map in a browser with a stylesheet that I created.

I’m using Docker for Windows and I first pulled ablondeau/tilemaker. It ran quickly and made a lot of mistakes.  Properties between polygons were transposed and the shapes of the polygons just weren’t right.

I then pulled osmtw/tilemaker which states it’s from this repository.  It took about 14 hours to run.  The final zoom level, 14 renders perfectly.  I found a missing polygon that wasn’t exceptionally complex but it was comprised of two outer polygons.  At zoom level 13 and lower, some of the tiles aren’t returned and the tileserver displays the error:
“Error: Vector Tile has POLYGON has ring with too few points. It is not valid according to v2 of VT spec.”

Any advice on resolving this?

Am I running the most recent version by using osmtw/tilemaker?

**Comment by systemed** [comment]
The most recent version is the one in this repository, systemed/tilemaker . The Docker build is not formally supported but you're welcome to try it and report any failures you encounter.

Do you have a link to the data you're trying?

**Comment by Lee858** [comment]
What is the correct syntax for installing the Docker image? 
I keep ending up with Repository & Tag = <none>.

**Comment by Lee858** [comment]
The attached is a small subset of the data that exhibits the problem. I made an mbtiles with the current build and there's a new error in addition to the previous.
"Error: Vector Tile has POLYGON with first ring clockwise. It is not valid according to v2 of VT spec."

[polygons.zip](https://github.com/systemed/tilemaker/files/4055667/polygons.zip)

**Comment by systemed** [comment]
Thanks - will take a look when I've got a moment.

**Comment by systemed** [comment]
Do you have a working MBGL style that I can test this with (to save having to reengineer one)?

**Comment by Lee858** [comment]
Attached.
[PolygonStyle.zip](https://github.com/systemed/tilemaker/files/4075036/PolygonStyle.zip)

**Comment by systemed** [comment]
I'm working through this but it's worth noting that you have several invalid geometries in your source data. For example:

- multipolygon relation 140126 has nested inner members (the golf course and the building within it)
- way 104007519 has a "spike" (i.e. it doubles back on itself)
- multipolygon relation 140095 has self-intersections

I think there are probably things we can do to make tilemaker more tolerant here, but I'd suggest that you try cleaning up the data first (not sure what you're using to create it?) which might help isolate what the tilemaker-specific issues actually are.

**Comment by systemed** [comment]
Tilemaker will now tell you about any invalid geometries in the source data. You can optionally be told about invalid geometries in the output by using the `--verbose` flag.

**Comment by Lee858** [comment]
I don't understand why the nested inner members are an issue when they're different objects that would be rendered separately. 

My primary problem seems to be polygons that touch themselves. Tilemaker is interpreting these as intersections.  The software that I'm using for editing and validation doesn't see touching as an intersection.  It seems to prefer touching polygons over creating additional nested polygons when subtracting one polygon from another.  I've attached a screenshot with two examples that can be achieved by touching or nesting.

![TouchingPolygons](https://user-images.githubusercontent.com/59800445/73601959-a0a9b280-451f-11ea-840c-97cfd222a45e.jpg)

**Comment by systemed** [comment]
Tilemaker uses Boost::Geometry internally and hence its definition of what’s a valid shape, which is based on the OGC standard. References:

https://www.boost.org/doc/libs/1_63_0/libs/geometry/doc/html/geometry/reference/algorithms/is_valid/is_valid_2_with_failure_value.html

https://www.boost.org/doc/libs/1_63_0/libs/geometry/doc/html/geometry/reference/enumerations/validity_failure_type.html

**Comment by Lee858** [comment]
Thanks for helping with this. I can write some code to nest the touching polygons.

**Comment by systemed** [comment]
No problem. Let me know if any further issues!

**Comment by Lee858** [comment]
Two requests:
1) Is it possible to provide more information about the invalid geometries to assist in locating them?
I'm dealing with government data that spans several thousand square miles and has hundreds of inner polygons.  Relation id is the only unique value being returned so it's difficult to locate the invalid geometry.  It would be helpful if it displayed the way or node ids.

2) A nice to have, would be to preface the verbose output of the invalid geometry with a common term (maybe "err:") that can be used to search the output. I'm piping the output to a text file so it ends up with all of the block and zoom text that is normally erased on screen.  Right now, I search for polygon because it's only in the output when invalid geometry exists.

**Comment by Lee858** [comment]
I used QGIS to identify problem data then wrote code to clean it up.  Tilemaker build no longer displays any errors that point to specific relations.  There are errors when writing tiles for each of the Zoom levels.   
"Output multipolygon has disconnected interior (contains polygons whose interiors are not disjoint)"

When I view the data using tileserver-mapnik, the logs display a number of errors with the tile directory and filename. Most are in in zoom levels 10 and 11.  I opened up some of the offending PBF tiles using QGIS then ran the Check Validity on them.  There are spikes and intersections from simplification. I checked the source data for the offending objects and they're valid.

I copied the offending objects to a separate file then created tiles from it. Tilemaker build doesn't display any errors.  If you open the PBF tiles with QGIS then run Check Validity, it shows the spikes and intersections.

Sample attached. I can provide more if needed.

Thanks


[SamplePolygons.zip](https://github.com/systemed/tilemaker/files/4397594/SamplePolygons.zip)
